File: test/config_reader/test_yaml_config_reader.cpp
#include <filesystem>

#include <gtest/gtest.h>
#include <mosaic/auto_configurer/config_reader/yaml_config_reader.h>

using namespace mosaic::auto_configurer;

class YamlConfigReaderTest : public ::testing::Test {
  protected:
    YamlConfigReader reader;
    std::string test_yaml_path;

    void SetUp() override {
        // Get the path to the test yaml file
        std::filesystem::path current_path = std::filesystem::current_path();
        test_yaml_path = (current_path / "test" / "config_reader" / "test_config.yaml").string();
    }
};

// Test IsSupportedExtension method
TEST_F(YamlConfigReaderTest, IsSupportedExtension_YamlExtension) {
    EXPECT_TRUE(reader.IsSupportedExtension("yaml"));
}

TEST_F(YamlConfigReaderTest, IsSupportedExtension_YmlExtension) {
    EXPECT_TRUE(reader.IsSupportedExtension("yml"));
}

TEST_F(YamlConfigReaderTest, IsSupportedExtension_UnsupportedExtension) {
    EXPECT_FALSE(reader.IsSupportedExtension("json"));
    EXPECT_FALSE(reader.IsSupportedExtension("xml"));
    EXPECT_FALSE(reader.IsSupportedExtension("txt"));
    EXPECT_FALSE(reader.IsSupportedExtension(""));
}

// Test LoadConfigs method
TEST_F(YamlConfigReaderTest, LoadConfigs_ValidYamlFile) {
    mosaic::core::MosaicConfig configs = reader.LoadConfigs(test_yaml_path);

    // Test server config
    EXPECT_EQ(configs.server.ws_url, "ws://localhost:8080");

    // Test auth config
    EXPECT_EQ(configs.server.auth_config.type, "basic");
    EXPECT_EQ(configs.server.auth_config.robot_id, "robot123");
    EXPECT_EQ(configs.server.auth_config.params.size(), 2);
    EXPECT_EQ(configs.server.auth_config.params["key1"], "value1");
    EXPECT_EQ(configs.server.auth_config.params["key2"], "value2");

    // Test WebRTC config
    EXPECT_EQ(configs.server.webrtc_config.ice_servers.size(), 2);

    // Test first ICE server
    EXPECT_EQ(configs.server.webrtc_config.ice_servers[0].urls.size(), 2);
    EXPECT_EQ(configs.server.webrtc_config.ice_servers[0].urls[0], "stun:stun.l.google.com:19302");
    EXPECT_EQ(configs.server.webrtc_config.ice_servers[0].urls[1], "stun:stun1.l.google.com:19302");
    EXPECT_EQ(configs.server.webrtc_config.ice_servers[0].username, "testuser");
    EXPECT_EQ(configs.server.webrtc_config.ice_servers[0].credential, "testcred");

    // Test second ICE server
    EXPECT_EQ(configs.server.webrtc_config.ice_servers[1].urls.size(), 1);
    EXPECT_EQ(configs.server.webrtc_config.ice_servers[1].urls[0], "turn:turn.example.com:3478");
    EXPECT_EQ(configs.server.webrtc_config.ice_servers[1].username, "turnuser");
    EXPECT_EQ(configs.server.webrtc_config.ice_servers[1].credential, "turncred");

    // Test connectors
    EXPECT_EQ(configs.connectors.size(), 2);

    // Test first connector
    EXPECT_EQ(configs.connectors[0].type, "camera");
    EXPECT_EQ(configs.connectors[0].label, "camera_dc");
    EXPECT_EQ(configs.connectors[0].params.size(), 2);
    EXPECT_EQ(configs.connectors[0].params["resolution"], "1920x1080");
    EXPECT_EQ(configs.connectors[0].params["fps"], "30");

    // Test second connector
    EXPECT_EQ(configs.connectors[1].type, "microphone");
    EXPECT_EQ(configs.connectors[1].label, "mic_dc");
    EXPECT_EQ(configs.connectors[1].params.size(), 2);
    EXPECT_EQ(configs.connectors[1].params["sample_rate"], "48000");
    EXPECT_EQ(configs.connectors[1].params["channels"], "2");
}

TEST_F(YamlConfigReaderTest, LoadConfigs_NonExistentFile) {
    EXPECT_THROW(reader.LoadConfigs("/non/existent/file.yaml"), std::exception);
}

File: python/test/tests/test_yaml_config_reader.py
"""
Equivalent of test/config_reader/test_yaml_config_reader.cpp

# NOTE: YamlConfigReader 및 IConfigReader는 현재 pybind11로 바인딩되지 않아
# 아래 테스트는 실행할 수 없습니다.  바인딩이 추가되면 주석을 제거하세요.
#
# 바인딩이 필요한 클래스:
#   - mosaic::config::IConfigReader        (인터페이스)
#   - mosaic::config::YamlConfigReader     (구현, IConfigReader 상속)
#
# 바인딩 완료 시 필요한 API:
#   YamlConfigReader()                        → 생성자
#   YamlConfigReader.is_supported_extension(ext: str) -> bool
#   YamlConfigReader.load_configs(path: str)  -> MosaicConfig
"""

# import unittest
# import mosaic_core


# class TestYamlConfigReaderIsSupportedExtension(unittest.TestCase):
#     """C++ TEST_F(YamlConfigReaderTest, IsSupportedExtension)"""
#
#     def setUp(self):
#         self.reader = mosaic_core.YamlConfigReader()
#
#     def test_yaml_supported(self):
#         self.assertTrue(self.reader.is_supported_extension(".yaml"))
#
#     def test_yml_supported(self):
#         self.assertTrue(self.reader.is_supported_extension(".yml"))
#
#     # C++ 테스트에서 지원하지 않는 확장자 목록
#     def test_json_not_supported(self):
#         self.assertFalse(self.reader.is_supported_extension(".json"))
#
#     def test_xml_not_supported(self):
#         self.assertFalse(self.reader.is_supported_extension(".xml"))
#
#     def test_txt_not_supported(self):
#         self.assertFalse(self.reader.is_supported_extension(".txt"))
#
#     def test_empty_not_supported(self):
#         self.assertFalse(self.reader.is_supported_extension(""))


# class TestYamlConfigReaderLoadConfigs(unittest.TestCase):
#     """C++ TEST_F(YamlConfigReaderTest, LoadConfigs)
#
#     C++ 테스트는 test/config_reader/test_config.yaml 파일을 읽어
#     MosaicConfig 구조체의 모든 필드를 검증합니다.
#     Python 테스트도 동일한 YAML 파일을 사용하여 검증합니다.
#     """
#
#     def setUp(self):
#         self.reader = mosaic_core.YamlConfigReader()
#         # C++ 테스트와 동일한 경로 (실행 CWD를 고려해 조정 필요)
#         self.config_path = "../../test/config_reader/test_config.yaml"
#
#     def test_load_configs(self):
#         config = self.reader.load_configs(self.config_path)
#
#         # server
#         self.assertEqual(config.server.ws_url, "wss://example.com")
#         # server.auth
#         self.assertEqual(config.server.auth_config.type, "token")
#         self.assertEqual(config.server.auth_config.robot_id, "robot-001")
#         self.assertEqual(config.server.auth_config.params, {"token": "abc123"})
#         # server.webrtc.ice_servers (2개)
#         self.assertEqual(len(config.server.webrtc_config.ice_servers), 2)
#         self.assertEqual(config.server.webrtc_config.ice_servers[0].urls, ["stun:stun.l.google.com:19302"])
#         self.assertEqual(config.server.webrtc_config.ice_servers[1].urls, ["turn:turn.example.com:3478"])
#         self.assertEqual(config.server.webrtc_config.ice_servers[1].username, "user")
#         self.assertEqual(config.server.webrtc_config.ice_servers[1].credential, "pass")
#         # connectors (2개)
#         self.assertEqual(len(config.connectors), 2)
#         self.assertEqual(config.connectors[0].type, "echo")
#         self.assertEqual(config.connectors[0].label, "echo_channel")
#         self.assertEqual(config.connectors[0].params, {"key1": "value1"})
#         self.assertEqual(config.connectors[1].type, "sensor")
#         self.assertEqual(config.connectors[1].label, "sensor_channel")
#         self.assertEqual(config.connectors[1].params, {"key2": "value2"})
#
#     def test_load_configs_file_not_found(self):
#         """C++ TEST_F(YamlConfigReaderTest, LoadConfigs) — non-existent file throws"""
#         with self.assertRaises(Exception):
#             self.reader.load_configs("nonexistent_file.yaml")


# if __name__ == "__main__":
#     unittest.main()